neurom/apps/morph_stats.py

```diff
@@ -73,34 +82,77 @@ def _stat_name(feat_name, stat_mode):
 
-def extract_stats(neurons, config):
+def extract_stats(neurons, config, as_frame=False):
+    """Extract stats from neurons.
+
+    Arguments:
+        neurons: a neuron, population or neurite tree
+        config (dict): configuration dict. The keys are:
+            - neurite_type: a list of neurite types for which features are extracted
+            - neurite: a dictionary {neurite_feature: mode} where:
+                - neurite_feature is a string from NEURITEFEATURES
+                - mode is an aggregation operation provided as a string such as:
+                  ['min', 'max', 'median', 'mean', 'std', 'raw', 'total']
+        as_frame (bool): if yes, returns a dataframe, else a dict
+
+    Returns:
+        The extracted statistics
+
+    Note:
+        An example config can be found at:
+    """
+    if as_frame:
+        return _extract_stats_as_frame(neurons, config)
+    return _extract_stats_as_dict(neurons, config)
+
+
+extract_stats.__doc__ = extract_stats.__doc__ + EXAMPLE_CONFIG + '\n'
+
+
+def _extract_stats_as_frame(neurons, config):
+    del config['neuron']
+    stats = {nrn.name: extract_stats(nrn, config) for nrn in neurons}
+    columns = list(next(iter(next(iter(stats.values())).values())).keys())
+
+    rows = [[name, neurite_type] + list(features.values())
+            for name, data in stats.items()
+            for neurite_type, features in data.items()]
+    return pd.DataFrame(columns=['name', 'neurite_type'] + columns,
+                        data=rows)
+
+
+def _extract_stats_as_dict(neurons, config):
     """Extract stats from neurons."""
-    stats = defaultdict(dict)
 
-    def _fill_compoundified(data, stat_name, stat):
-        """Insert the stat in the dict and eventually split it into XYZ components."""
+    def _fill_stats_dict(data, stat_name, stat):
+        """Insert the stat data in the dict.
+
+        And if the stats is a 3D array, splits it into XYZ components.
+        """
         if stat is None or not isinstance(stat, np.ndarray) or stat.shape not in ((3, ), ):
             data[stat_name] = stat
         else:
             for i, suffix in enumerate('XYZ'):
                 compound_stat_name = stat_name + '_' + suffix
                 data[compound_stat_name] = stat[i]
 
+    stats = defaultdict(dict)
+
     for (feature_name, modes), neurite_type in product(config['neurite'].items(),
                                                        config['neurite_type']):
         neurite_type = _NEURITE_MAP[neurite_type]
         for mode in modes:
             stat_name = _stat_name(feature_name, mode)
             stat = eval_stats(nm.get(feature_name, neurons, neurite_type=neurite_type), mode)
-            _fill_compoundified(stats[neurite_type.name], stat_name, stat)
+            _fill_stats_dict(stats[neurite_type.name], stat_name, stat)
 
-    for feature_name, modes in config['neuron'].items():
+    for feature_name, modes in config.get('neuron', {}).items():
         for mode in modes:
```

Comment:
in the config you mentionned in the discussion, there is a single mode not in a list, so it crashed here as well. Maybe we can add
```
if not isinstance(modes, list):
    modes = [modes]
```

Reply:
I don't get what is crashing. This seems to work:
```python
    nrns = nm.load_neurons([os.path.join(SWC_PATH, name)
                            for name in ['Neuron.swc', 'simple.swc']])
    config = {'neurite': {'total_length_per_neurite': ['total']},
              'neurite_type': ['AXON', 'APICAL_DENDRITE', 'BASAL_DENDRITE']}
    actual = ms.extract_dataframe(nrns, config)
```
Can you elaborate ?

Reply:
I use that config:
```
CONFIG_TOTAL_LENGTH = {
    "neurite": {"total_length_per_neurite": "total"},
    "neurite_type": {"AXON", "APICAL_DENDRITE", "BASAL_DENDRITE"},
}
```

Reply:
I use that config:
```
CONFIG_TOTAL_LENGTH = {
    "neurite": {"total_length_per_neurite": "total"},
    "neurite_type": {"AXON", "APICAL_DENDRITE", "BASAL_DENDRITE"},
}
```

Reply:
But this is fine, we can pass a list, but I thought it may be more user friendly to be able to pass a single string

Reply:
I'm not a huge fan either of this config format but since it is already in use in NeuroM we cannot really change it.
However I have made the "neurite_type" key optional.